File: 프로그래머스/1/42840. 모의고사/모의고사.py
'''
1.아이디어
    - 수포자 3인방의 찍는 가장 간단한 part를 따로 저장한하여, answers가 주어지면 해당 순서에 맞게 반복해서 비교한다
    - 맞은 개수를 저장하고, 가장 많은 문제를 맞힌 사람을 배열에 담아 반환/여러 명 가능
'''

def solution(answers):
    # 수포자들이 찍는 규칙
    std1 = [1, 2, 3, 4, 5]
    std2 = [2, 1, 2, 3, 2, 4, 2, 5]
    std3 = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    
    # 맞춘 개수 카운트 초기화
    cnt1 = 0
    cnt2 = 0
    cnt3 = 0 
    
    # 각 수포자들이 찍은 답과 answers를 비교해서 맞춘 개수 세기
    for idx, answer in enumerate(answers):
        if std1[idx % len(std1)] == answer:
            cnt1 += 1
        if std2[idx % len(std2)] == answer:
            cnt2 += 1
        if std3[idx % len(std3)] == answer:
            cnt3 += 1
    
    # 각 수포자들의 점수 리스트
    std_scores = [cnt1, cnt2, cnt3]
    max_score = max(std_scores)  # 가장 많이 맞춘 점수를 구함
    
    # 가장 많은 점수를 맞춘 사람들의 번호 반환
    return [idx + 1 for idx, score in enumerate(std_scores) if score == max_score]

File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.*;

class Solution {
    public int[] solution(int[] answers) {
        int n = answers.length; // 시험 문제 수
        int[] scores = new int[4]; // 학생 점수 초기화
        for(int i=1; i<=3; i++){
            scores[i] = 0;
        }
        
        ArrayList<Integer>[] std = new ArrayList[4];
        for(int i=1; i<=3; i++){
            std[i] = new ArrayList<>();
        }
        std[1].add(1); std[1].add(2); std[1].add(3); std[1].add(4); std[1].add(5);
        std[2].add(2); std[2].add(1);
        std[2].add(2); std[2].add(3);
        std[2].add(2); std[2].add(4);
        std[2].add(2); std[2].add(5);
        std[3].add(3); std[3].add(3);
        std[3].add(1); std[3].add(1);
        std[3].add(2); std[3].add(2);
        std[3].add(4); std[3].add(4);
        std[3].add(5); std[3].add(5);
        
        
        for(int i=1; i<=3; i++){
            for(int j=0; j<n; j++){
                int num = std[i].get(j%(std[i].size()));
                if(num == answers[j]) scores[i] += 1;
            }
        }
        
        
        int maxVal = -1;
        for(int i=1; i<4; i++){
            maxVal = Math.max(maxVal, scores[i]);
        }
        
        List<Integer> result = new ArrayList<>();
        for(int i=1; i<4; i++){
            if(scores[i] == maxVal) result.add(i);
        }
        
        int[] answer = new int[result.size()];
        for(int i=0; i<result.size(); i++){
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}